brambling/views/orders.py

```diff
@@ -306,11 +293,42 @@ def get_reverse_kwargs(self):
             'event_slug': self.event.slug,
             'organization_slug': self.event.organization.slug,
         }
-        if self.kwargs.get('code'):
-            kwargs['code'] = self.order.code
         return kwargs
 
 
+class OrderCodeRedirectView(OrderMixin, View):
+
+    def dispatch(self, request, *args, **kwargs):
+        self.event = get_object_or_404(
+            Event.objects.select_related('organization'),
+            slug=kwargs['event_slug'],
+            organization__slug=kwargs['organization_slug'])
+        if not self.event.viewable_by(request.user):
+            raise Http404
+        try:
+            order = Order.objects.get(event=self.event, code=kwargs['code'])
+        except Order.DoesNotExist:
+            raise Http404
+
+        session_code = Order.objects._get_session_code(request, self.event)
+        order_in_session = kwargs['code'] == session_code
+        if (order and not order.person and not request.user.is_authenticated()
+            and not order_in_session):
+            url = reverse('brambling_signup')
+            messages.add_message(request, messages.ERROR, "Please sign up with an account to view this order.")
```

Comment:
Since we're redirecting, might be good to add the order code into this message & below. Something like: "Order abcdef was created without an account. Create an account with the same email address as the order to view it."